### DISCUSSION TITLE
General: Why enterprise level project use Spring Boot when it have low performance?

### DISCUSSION THREAD

**Discussion by indramal** [TechEmpower/FrameworkBenchmarks#8168] [Q&A]
![image](https://user-images.githubusercontent.com/28696256/233769017-516ec15a-5394-421a-82de-d3698ceeea33.png)
![image](https://user-images.githubusercontent.com/28696256/233769024-9023047b-fd5b-4bc4-98b5-9595da5cefc3.png)

Considering popular languages, framwork and servers, high performance languages are PHP, Go after that NodeJS. Why some of developers use Spring Boot when have low performance?

I am asking this in performance and features.

**Comment by remittor** [comment]
> Why some of developers use Spring Boot when have low performance?

Look this: https://github.com/TechEmpower/FrameworkBenchmarks/wiki/TechEmpower-Performance-Rating-(TPR)#tentative-selections

In the `Python` world, the situation is similar with `Django` (it is marked with the `TPR` flag and at the same time it is very slow).
Yes, it's strange, but such is our world...

**Reply by indramal** [reply]
Some blogs says spring boot have lot of tools for productions then time is less for productions.

**Comment by ishtms** [comment]
You shouldn't really take the numbers in benchmarks too seriously unless you've built a server which is already affected by slowed down performance. Spring still has 20k requests per second, which is more than enough to handle any kind of load. If you're receiving 20k requests per second for at-least 3 hours every day that amounts to around **215 million requests**. And if your application is serving that much traffic, you should already be in a state to financially support multiple servers and scale horizontally, or maybe vertically if your server's CPU or Memory is being bottleneck'ed.

**Reply by indramal** [reply]
> You shouldn't really take the numbers in benchmarks too seriously unless you've built a server which is already affected by slowed down performance. 

If you build large website backend, then it takes much time to convert low resources use language. It is cost. If not convert, need to buy another server to scale. That also cost. Why can't we use low resources use language from beginning (at backend building stage) ? 

You are correct, but we can reduce no of servers when consider performance.  Spring Boot need high resources and it is cost. When consider scaling, need to scale in low traffic compare to others also need high performance server than PHP needs. 

So considering costing,

PHP can low resources server, Spring Boot need more resources server - High cost for Spring Boot
Firstly, Spring Boot needs to scale compare to PHP to get the same level of performance becouse Spring Boot need more resources server. -  High cost for Spring Boot

Why not use PHP or Rust or Go language instead of NodeJS and Spring Boot?

**Comment by re-thc** [comment]
Spring Boot doesn't really have low performance and you're comparing apples to oranges.

Why do you compare PHP (raw) to Spring Boot (framework)? At least compare Laravel (framework) to Spring Boot (framework). If you compare the language Java vs PHP, Java is top of the list...

As to the tests - certain frameworks / tests are better optimized for the benchmark. Spring isn't 1 of them. You can compare the Spring code in this repository vs e.g. Vertx and you'll see the differences.

In the real world, Spring is slower but still much closer to Quarkus, Helidon, Micronaut, etc (the other Java frameworks) compared to where it sits in TechEmpower.

**Comment by schinraw** [comment]
These benchmarks do not make any sense, for Just they have used Caching mechanism for single query while in spring they are making call to database. Similarly for update operations Just is using relational database with direct update query while spring framework using MongoDb with JPA update, which will first make a read call and if row is present then make update call. There is no way javascript or php beating spring boot applications
